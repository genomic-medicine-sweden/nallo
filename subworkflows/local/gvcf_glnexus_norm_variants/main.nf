//
// Takes a channel of grouped gVCFs, merged them with GLNexus,
// adds FOUND_IN tag, normalizes and decomposes variants.
//
include { ADD_FOUND_IN_TAG                           } from '../../../modules/local/add_found_in_tag/main'
include { BCFTOOLS_NORM as BCFTOOLS_NORM_MULTISAMPLE } from '../../../modules/nf-core/bcftools/norm/main'
include { GLNEXUS                                    } from '../../../modules/nf-core/glnexus/main'

workflow GVCF_GLNEXUS_NORM_VARIANTS {
    take:
    ch_gvcfs       // channel: [mandatory] [ val(meta), path(gvcfs)]
    ch_bed         // channel: [optional]  [ val(meta), path(input_bed) ]
    ch_fasta       // channel: [mandatory] [ val(meta), path(fasta) ]
    variant_caller // string: variant caller to tag the variants with, e.g. "deepvariant"

    main:
    ch_versions = Channel.empty()

    GLNEXUS(
        ch_gvcfs,
        ch_bed,
    )
    ch_versions = ch_versions.mix(GLNEXUS.out.versions)

    // Annotate with FOUND_IN tag - not sure what would happen if we do this before glnexus instead?
    ADD_FOUND_IN_TAG(
        GLNEXUS.out.bcf.map { meta, vcf -> [meta, vcf, []] },
        variant_caller,
    )
    ch_versions = ch_versions.mix(ADD_FOUND_IN_TAG.out.versions)

    // Decompose and normalize variants
    BCFTOOLS_NORM_MULTISAMPLE(
        ADD_FOUND_IN_TAG.out.vcf.map { meta, vcf -> [meta, vcf, []] },
        ch_fasta,
    )
    ch_versions = ch_versions.mix(BCFTOOLS_NORM_MULTISAMPLE.out.versions)

    emit:
    vcf      = BCFTOOLS_NORM_MULTISAMPLE.out.vcf                                        // channel: [ val(meta), path(vcf) ]
    index    = BCFTOOLS_NORM_MULTISAMPLE.out.tbi.mix(BCFTOOLS_NORM_MULTISAMPLE.out.csi) // channel: [ val(meta), path(tbi/csi) ]
    versions = ch_versions                                                              // channel: [ path(versions.yml) ]
}
