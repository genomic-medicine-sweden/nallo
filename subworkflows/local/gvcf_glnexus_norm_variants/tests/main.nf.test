nextflow_workflow {

    name "Test Workflow GVCF_GLNEXUS_NORM_VARIANTS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "GVCF_GLNEXUS_NORM_VARIANTS"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }

        }
    }

    test("1 gVCF, without bed") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant.g.vcf.gz', checkIfExists: true),
                ])
                input[1] = [[],[]]
                input[2] = GUNZIP.out.gunzip
                input[3] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("2 gVCFs, without bed") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant.g.vcf.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/HG003_deepvariant.g.vcf.gz', checkIfExists: true),
                    ]
                ])
                input[1] = [[],[]]
                input[2] = GUNZIP.out.gunzip
                input[3] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("2 gVCFs, with bed") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant.g.vcf.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/HG003_deepvariant.g.vcf.gz', checkIfExists: true),
                    ]
                ])
                input[1] = [
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'reference/test_data.bed', checkIfExists: true),
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("1 gVCF, without bed -stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant.g.vcf.gz', checkIfExists: true),
                ])
                input[1] = [[],[]]
                input[2] = GUNZIP.out.gunzip
                input[3] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("2 gVCFs, without bed -stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant.g.vcf.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/HG003_deepvariant.g.vcf.gz', checkIfExists: true),
                    ]
                ])
                input[1] = [[],[]]
                input[2] = GUNZIP.out.gunzip
                input[3] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

    test("2 gVCFs, with bed -stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/HG002_deepvariant.g.vcf.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/HG003_deepvariant.g.vcf.gz', checkIfExists: true),
                    ]
                ])
                input[1] = [
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'reference/test_data.bed', checkIfExists: true),
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = 'deepvariant'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                ).match() }
            )
        }

    }

}
