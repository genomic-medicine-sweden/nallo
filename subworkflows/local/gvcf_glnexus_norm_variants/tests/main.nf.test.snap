{
    "2 gVCFs, with bed -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:04:52.235260558"
    },
    "1 gVCF, without bed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,55f3c68656e8c179bb1fe61057ce1bc3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,f4abe98ed0cb3c2b428c26ad0635d72e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,f4abe98ed0cb3c2b428c26ad0635d72e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,55f3c68656e8c179bb1fe61057ce1bc3"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:04:20.181457758"
    },
    "2 gVCFs, without bed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,64386cbe4844c9ec0a7aa78c82ff8ab1"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,0a0f558ffb70dfeb7ba6220a9cfb6b0d"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,0a0f558ffb70dfeb7ba6220a9cfb6b0d"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,64386cbe4844c9ec0a7aa78c82ff8ab1"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:04:26.908365122"
    },
    "2 gVCFs, without bed -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:04:46.054196423"
    },
    "1 gVCF, without bed -stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:04:39.703473629"
    },
    "2 gVCFs, with bed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,64386cbe4844c9ec0a7aa78c82ff8ab1"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,0a0f558ffb70dfeb7ba6220a9cfb6b0d"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ],
                "index": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz.tbi:md5,0a0f558ffb70dfeb7ba6220a9cfb6b0d"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf.gz:md5,64386cbe4844c9ec0a7aa78c82ff8ab1"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38ea33a2f4cd8fe61103a16e2de83f53",
                    "versions.yml:md5,4ddf96fdce24e35aebe4ace35857591b",
                    "versions.yml:md5,fef01d57d7aab64c1458c63ffe18d1a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-10-22T11:04:33.439816913"
    }
}